pandas/core/indexes/timedeltas.py

```diff
@@ -154,32 +153,73 @@ def __new__(cls, data=None, unit=None, freq=None, start=None, end=None,
+        if isinstance(data, TimedeltaIndex) and freq is None and name is None:
+            if copy:
+                return data.copy()
+            else:
+                return data._shallow_copy()
+
+        # - Cases checked above all return/raise before reaching here - #
+
+        if unit is not None:
             data = to_timedelta(data, unit=unit, box=False)
-        elif copy:
-            data = np.array(data, copy=True)
-
-        data = np.array(data, copy=False)
-        if data.dtype == np.object_:
-            data = array_to_timedelta64(data)
-        if data.dtype != _TD_DTYPE:
-            if is_timedelta64_dtype(data):
+
+        # Unwrap whatever we have into a np.ndarray
+        if not hasattr(data, 'dtype'):
+            # e.g. list, tuple
+            if np.ndim(data) == 0:
+                # i.e.g generator
+                data = list(data)
+            data = np.array(data, copy=False)
+        elif isinstance(data, ABCSeries):
+            data = data._values
+        elif isinstance(data, (cls, TimedeltaArrayMixin)):
+            data = data._data
+
+        # Convert whatever we have into timedelta64[ns] dtype
+        if is_object_dtype(data) or is_string_dtype(data):
+            # no need to make a copy, need to convert if string-dtyped
+            data = np.array(data, dtype=np.object_, copy=False)
+            data = array_to_timedelta64(data).view(_TD_DTYPE)
+            copy = False
+        elif is_integer_dtype(data):
+            # treat as nanoseconds
+            # if something other than int64, convert
+            data = ensure_int64(data)
+            if copy:
+                # TODO: can we avoid branching here? `astype(data, copy=False)`
+                #  appears to be making a copy
+                data = data.astype(_TD_DTYPE)
+                copy = False
+            else:
+                data = data.view(_TD_DTYPE)
+        elif is_float_dtype(data):
+            # We allow it if and only if it can be converted lossessly
+            mask = np.isnan(data)
+            casted = data.astype(np.int64)
+            if not (casted[~mask] == data[~mask]).all():
+                raise TypeError("floating-dtype data cannot be losslessly "
+                                "converted to {cls}".format(cls=cls.__name__))
+            data = casted.view(_TD_DTYPE)
+            data[mask] = iNaT
+            copy = False
+        elif is_timedelta64_dtype(data):
+            if data.dtype != _TD_DTYPE:
                 # non-nano unit
                 # TODO: watch out for overflows
                 data = data.astype(_TD_DTYPE)
-            else:
-                data = ensure_int64(data).view(_TD_DTYPE)
+                copy = False
+        else:
+            raise TypeError("dtype {dtype} is invalid for constructing {cls}"
+                            .format(dtype=data.dtype, cls=cls.__name__))
 
-        assert data.dtype == 'm8[ns]', data.dtype
+        data = np.array(data, copy=copy)
```

Comment:
Can we use np.asarray here?

Reply:
same as above, this needs to be specifically an np.ndarray